File: src/main/java/dragonbook/inter/expr/Arith.java
package dragonbook.inter.expr;

import dragonbook.error.ParseError;
import dragonbook.lexer.Token;
import dragonbook.symbols.Type;

public record Arith(Token op, Expr expr1, Expr expr2) implements Op {

    public Arith {
        if (!expr1.type().isNumeric() || !expr2.type().isNumeric()) throw new ParseError("type error");
    }

    @Override
    public Type type() {
        return max(expr1.type(), expr2.type());
    }

    private static Type max(Type type1, Type type2) {
        if (type1 == Type.FLOAT || type2 == Type.FLOAT) return Type.FLOAT;
        else if (type1 == Type.INT || type2 == Type.INT) return Type.INT;
        return Type.CHAR;
    }
}

File: src/main/kotlin/dragonbook/inter/expr/Arith.kt
package dragonbook.inter.expr

import dragonbook.error.ParseError
import dragonbook.lexer.Token
import dragonbook.symbols.Type

data class Arith(
    override val op: Token,
    val expr1: Expr,
    val expr2: Expr
) : Op {

    init {
        validate()
    }

    private fun validate() {
        if (!expr1.type.isNumeric() || !expr2.type.isNumeric()) throw ParseError(node = this)
    }

    override val type: Type
        get() =
            if (expr1.type == Type.FLOAT || expr2.type == Type.FLOAT) Type.FLOAT
            else if (expr1.type == Type.INT || expr2.type == Type.INT) Type.INT
            else Type.CHAR

    override fun toString(): String = "$expr1 $op $expr2"
}